### PR TITLE
Fixes #2828 - Report an error when trying to create/update ovirt resource

### PR DESCRIPTION
**Pull request by ik5**
On both create and update, there was no exception handler to captured
given exception. That resulted in having an exception screen instead of
an error message.
The fix capture all of the given exceptions, and report them as an error.

### CODE REVIEW

**Comment by tbrisker** [comment]
ok to test

File: app/controllers/compute_resources_controller.rb
Comment (adamruzicka) [inline app/controllers/compute_resources_controller.rb]: i think the `% e.message` should go outside the `_()` translation block. Also it would be probably nice to log the full backtrace of the exception somewhere in case we would need to debug it.

**Review by adamruzicka — COMMENTED** [review]
Works as expected for me, left one comment inline.

File: app/controllers/compute_resources_controller.rb
Comment (tbrisker) [inline app/controllers/compute_resources_controller.rb]: This isn't necessarily an ovirt resource

File: app/controllers/compute_resources_controller.rb
Comment (tbrisker) [inline app/controllers/compute_resources_controller.rb]: This isn't necessarily an ovirt resource

**Review by tbrisker — CHANGES_REQUESTED** [review]
a couple of minor comments inline. Also, is there some more specific exception we expect to occur here that isn't `StandardError`? I would prefer rescuing only what we expect to be raised if possible. (e.g. this would catch exceptions which we already handle differently in application controller)

File: app/controllers/compute_resources_controller.rb
Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: Right. when an exception comes back to the controller, I do not want it to return an exception screen, or 500 page, but to display an error message to the user.

If an exception was not handled prior to that location, at least allow the user to continue doing other stuff.

File: app/controllers/compute_resources_controller.rb
Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: see my comment above

File: app/controllers/compute_resources_controller.rb
Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: @tbrisker was probably referring to the `ovirt` word in `Error while creating ovirt resource'`, it should be `Error while creating compute resource`

Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: waiting for a PR of mine about unified Ovirt exception, and then I'll change this change accordingly.

Comment (boaz0) [reply app/controllers/compute_resources_controller.rb]: I agree with @tbrisker.
Lets keep the error message general and in the next PR where ovirt exception is thrown, add a more explicit error message saying it was `ovirt`.

WDYT?

Comment (tbrisker) [reply app/controllers/compute_resources_controller.rb]: I was indeed referring to the error message here. This is the compute resource controller which is used by all compute resources. Any error message should not contain a reference to ovirt, since that won't make sense, for example, when someone hits an error when trying to create a vmware resource.

Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: sorry for that, you are right, and now added support for it.

File: app/controllers/compute_resources_controller.rb
Comment (iNecas) [inline app/controllers/compute_resources_controller.rb]: There is a logger helper that can/should be used, see for example https://github.com/theforeman/foreman/blob/4a96211d33c261655c887bbb6c787e38d883200a/app/validators/ssh_key_validator.rb#L13.

Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: Also, there is ovirt mentioned here as well, while it should not (as well as in other places of this PR)

Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: I'm missing here something in my understanding.
The original issue was, that when exception was raised, the actual 500 page would have shown rather then an error message the "foreman way".

My original attempt was to make it display the "foreman way", so if I create a validator here, how would it solve the problem of having the 500 error page?

Having said that, I do understand that the controller is agnostic to the worker that doing the actual task, and will solve this.

Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: No, I haven't meant validator: it was just an example of logging the exception with the `Logging.exception` API

Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: Btw. since we rescue `StandardError` at the other place, I wonder, if we shouldn't use this here as well compared to `Fog::Ovirt::Errors::OvirtEngineError`

Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: The original code used `StandardError`, and moved it to use `Fog::Ovirt::Errors::OvirtEngineError` following the above discussion,

Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: I'm ok with that then. Still asking for using `Foreman::Logging.exception` use here

File: app/controllers/compute_resources_controller.rb
Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: We are rescuing from ` StandardError => e` which is not ovirt specific, therefore the message should not have `ovirt`, Am I reading this right @tbrisker?

Comment (tbrisker) [reply app/controllers/compute_resources_controller.rb]: exactly, since the controller is generic for all compute resources, if for example this is a libvirt resource it doesn't make sense to log an error referring to ovirt.

Comment (tbrisker) [reply app/controllers/compute_resources_controller.rb]: but shouldn't this also only catch ovirt errors instead of any generic ones like the other method?

**Comment by ik5** [comment]
please note that there is a need to wait for https://github.com/fog/fog-ovirt/pull/8 to be merged (by @orrabin ) and bumped a version before merging this PR

**Comment by orrabin** [comment]
@ik5 that will take time, I suggest to use the existing exception and later fix this as part of the fix we'll need to do in core for the new fog-ovirt version.

**Comment by ik5** [comment]
following @orrabin comment, I moved my code to use `OVIRT::OvirtException` instead of `Fog::Ovirt::Errors::OvirtEngineError`. i will need to change it once the new ovirt version will be released.

File: app/controllers/compute_resources_controller.rb
Comment (tbrisker) [inline app/controllers/compute_resources_controller.rb]: This is update, not create

File: app/controllers/compute_resources_controller.rb
Comment (tbrisker) [inline app/controllers/compute_resources_controller.rb]: Why do we catch `StandardError` here and `OVIRT::OvirtException` above? I expect that we should behave the same in both cases.
And if we want to catch a more generic error class, I would go with `Fog::Errors::Error` perhaps which fog-ovirt should inherit for its errors?

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: app/controllers/compute_resources_controller.rb
Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: fixed

File: app/controllers/compute_resources_controller.rb
Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: Okay, following our conversation f2f, I'ved also added a "fall-back" so both `Fog::Error::Error` and `OVIRT::OvirtException` will be captured.

File: app/controllers/compute_resources_controller.rb
Comment (tbrisker) [inline app/controllers/compute_resources_controller.rb]: nitpick - this should be `...to update resource`

**Review by tbrisker — CHANGES_REQUESTED** [review]
one tiny nitpick, other than that LGTM

File: app/controllers/compute_resources_controller.rb
Comment (ik5) [reply app/controllers/compute_resources_controller.rb]: fixed, thanks :)

**Review by tbrisker — APPROVED** [review]
LGTM, ready to merge once tests pass.

File: app/controllers/compute_resources_controller.rb
Comment (ohadlevy) [inline app/controllers/compute_resources_controller.rb:39]: should we add some sort of TODO here? as long term I dont expect to see rbovirt errors in our codebase? (or in other words, I would expect that after @orrabin finish the fog-ovirt rewrite, we will not know rbovirt in foreman code base at all).

Comment (tbrisker) [reply app/controllers/compute_resources_controller.rb:39]: @ohadlevy this isn't the only place we reference `OVIRT` module or exceptions. Ideally, we shouldn't care about the underlying implementation here and for other resource providers and push anything that does care into fog (I think we also refer to some RbVmomi stuff elsewhere).
Might be a good idea to open a separate issue to remove all references once `fog-ovirt` handles them. What do you think @orrabin ?

Comment (orrabin) [reply app/controllers/compute_resources_controller.rb:39]: That is the plan. Once we start using the version of `fog-ovirt` that wraps exceptions we will change all references to `OVIRT::OvirtException` in foreman. There isn't an issue for it yet, I'll open it.

**Comment by tbrisker** [comment]
Thanks @ik5 !

**Pull request merged**
